presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayToArrayCast.java

```diff
@@ -0,0 +1,131 @@
+import com.facebook.presto.sql.gen.ArrayGeneratorUtils;
+import com.facebook.presto.sql.gen.ArrayMapByteCodeExpression;
+import com.facebook.presto.sql.gen.CallSiteBinder;
+import com.facebook.presto.sql.gen.CompilerUtils;
+import com.facebook.presto.type.ArrayType;
+import com.facebook.presto.type.TypeUtils;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.Slice;
+
+import java.lang.invoke.MethodHandle;
+import java.util.Map;
+
+import static com.facebook.presto.byteCode.Access.FINAL;
+import static com.facebook.presto.byteCode.Access.PRIVATE;
+import static com.facebook.presto.byteCode.Access.PUBLIC;
+import static com.facebook.presto.byteCode.Access.STATIC;
+import static com.facebook.presto.byteCode.Access.a;
+import static com.facebook.presto.byteCode.Parameter.arg;
+import static com.facebook.presto.byteCode.ParameterizedType.type;
+import static com.facebook.presto.byteCode.expression.ByteCodeExpressions.constantBoolean;
+import static com.facebook.presto.byteCode.expression.ByteCodeExpressions.invokeStatic;
+import static com.facebook.presto.metadata.FunctionRegistry.operatorInfo;
+import static com.facebook.presto.metadata.OperatorType.CAST;
+import static com.facebook.presto.metadata.Signature.internalOperator;
+import static com.facebook.presto.metadata.Signature.typeParameter;
+import static com.facebook.presto.spi.StandardErrorCode.FUNCTION_NOT_FOUND;
+import static com.facebook.presto.sql.gen.CompilerUtils.defineClass;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+
+public class ArrayToArrayCast
+        extends ParametricOperator
+{
+    public static final ArrayToArrayCast ARRAY_TO_ARRAY_CAST = new ArrayToArrayCast();
+
+    private ArrayToArrayCast()
+    {
+        super(CAST, ImmutableList.of(typeParameter("F"), typeParameter("T")), "array<T>", ImmutableList.of("array<F>"));
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        checkArgument(arity == 1, "Expected arity to be 1");
+        Type fromType = types.get("F");
+        Type toType = types.get("T");
+
+        ArrayType fromArrayType = (ArrayType) typeManager.getParameterizedType(StandardTypes.ARRAY, ImmutableList.of(fromType.getTypeSignature()), ImmutableList.of());
+        ArrayType toArrayType = (ArrayType) typeManager.getParameterizedType(StandardTypes.ARRAY, ImmutableList.of(toType.getTypeSignature()), ImmutableList.of());
+
+        FunctionInfo functionInfo = functionRegistry.getExactFunction(internalOperator(CAST.name(), toType.getTypeSignature(), ImmutableList.of(fromType.getTypeSignature())));
+        if (functionInfo == null) {
+            throw new PrestoException(FUNCTION_NOT_FOUND, format("Can not cast %s to %s", fromArrayType, toArrayType));
```

Comment:
Instead of throwing an exception here, we should introduce "function parameters" for unbound signatures. These would work like type parameters, but bind to functions instead.